### PR TITLE
Add separate const offset from gep pass to SyncVM

### PR DESCRIPTION
**Pull request by akiramenai**
Update tests as the pass asserts no unused instruction are there if
`--reassociate-geps-verify-no-dead-code` is specified.

### CODE REVIEW

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.083 ║
║ Best                               21.138 ║
║ Worst                              -1.070 ║
║ Total                               0.159 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.020 ║
║ Best                               21.782 ║
║ Worst                              -7.085 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.011 ║
║ Best                               10.447 ║
║ Worst                              -6.210 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.555 ║
║ Best                                5.243 ║
║ Worst                               0.000 ║
║ Total                               0.442 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.005 ║
║ Best                                0.652 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.002 ║
║ Best                                0.301 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.083 ║
║ Best                               21.138 ║
║ Worst                              -1.070 ║
║ Total                               0.159 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.020 ║
║ Best                               21.782 ║
║ Worst                              -7.085 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.011 ║
║ Best                               10.447 ║
║ Worst                              -6.210 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.555 ║
║ Best                                5.243 ║
║ Worst                               0.000 ║
║ Total                               0.442 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.005 ║
║ Best                                0.652 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.002 ║
║ Best                                0.301 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╚═══════════════════════════════════════════╝

```

File: llvm/test/CodeGen/SyncVM/copytoreg_bug.ll [hunk 1]
```diff
@@ -1,16 +0,0 @@
-; RUN: llc -O0 < %s | FileCheck %s
```
Comment (akiramenai) [inline llvm/test/CodeGen/SyncVM/copytoreg_bug.ll:1]: This test is ISA 1.0 legacy, it asserted that we don't ICE in codegen. Removed it as it's no longer needed

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.083 ║
║ Best                               21.138 ║
║ Worst                              -1.070 ║
║ Total                               0.159 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.020 ║
║ Best                               21.782 ║
║ Worst                              -7.085 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.011 ║
║ Best                               10.447 ║
║ Worst                              -6.210 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.555 ║
║ Best                                5.243 ║
║ Worst                               0.000 ║
║ Total                               0.442 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.005 ║
║ Best                                0.652 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.002 ║
║ Best                                0.301 ║
║ Worst                               0.000 ║
║ Total                               0.001 ║
╚═══════════════════════════════════════════╝

```

**Review by PavelKopyl — APPROVED** [review]
LGTM.

It could also be interesting to check the following pipeline (see: NVPTXPassConfig::addStraightLineScalarOptimizationPasses()):
```
addPass(createSeparateConstOffsetFromGEPPass());
 // ReassociateGEPs exposes more opportunites for SLSR. See
 // the example in reassociate-geps-and-slsr.ll.
 addPass(createStraightLineStrengthReducePass());
 // SeparateConstOffsetFromGEP and SLSR creates common expressions which GVN or
 // EarlyCSE can reuse. GVN generates significantly better code than EarlyCSE
 // for some of our benchmarks.
 addPass(createGVNPass());
 // Run NaryReassociate after EarlyCSE/GVN to be more effective.
 addPass(createNaryReassociatePass());
 // NaryReassociate on GEPs creates redundant common expressions, so run
 // EarlyCSE after it.
 addPass(createEarlyCSEPass());
```

**Comment by akiramenai** [comment]
@PavelKopyl will check. Thanks!

**Pull request merged**